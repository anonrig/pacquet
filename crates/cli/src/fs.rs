use std::{io, os, path::Path};

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::unix::fs::symlink(original, link)
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::windows::fs::symlink_dir(original, link)
}

#[cfg(test)]
pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

#[cfg(test)]
pub fn get_all_folders(root: &std::path::PathBuf) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path().to_path_buf();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_string_lossy().to_string())
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path.as_os_string());
            }
        }
    }
    files.sort();
    files
}


