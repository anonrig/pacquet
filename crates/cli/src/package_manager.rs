use std::{io, path::PathBuf};

use pacquet_diagnostics::{
    miette::{self, Diagnostic},
    thiserror::{self, Error},
};
use pacquet_npmrc::{current_npmrc, Npmrc};
use pacquet_package_json::PackageJson;

use crate::package_cache::PackageCache;

#[derive(Error, Debug, Diagnostic)]
pub enum AutoImportError {
    #[error("cannot create directory at {dirname:?}: {error}")]
    CreateDir {
        dirname: PathBuf,
        #[source]
        error: io::Error,
    },
    #[error("fail to create a link from {from:?} to {to:?}: {error}")]
    CreateLink {
        from: PathBuf,
        to: PathBuf,
        #[source]
        error: io::Error,
    },
}

#[derive(Error, Debug, Diagnostic)]
#[non_exhaustive]
pub enum PackageManagerError {
    #[error(transparent)]
    #[diagnostic(transparent)]
    Tarball(#[from] pacquet_tarball::TarballError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    PackageJson(#[from] pacquet_package_json::PackageJsonError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    Registry(#[from] pacquet_registry::RegistryError),

    #[error(transparent)]
    #[diagnostic(code(pacquet_package_manager::io_error))]
    Io(#[from] io::Error),

    #[error(transparent)]
    #[diagnostic(transparent)]
    AutoImport(#[from] AutoImportError),
}

pub struct PackageManager {
    pub config: Box<Npmrc>,
    pub package_json: Box<PackageJson>,
    pub http_client: Box<reqwest::Client>,
    pub(crate) package_cache: PackageCache,
}

impl PackageManager {
    pub fn new<P: Into<PathBuf>>(package_json_path: P) -> Result<Self, PackageManagerError> {
        Ok(PackageManager {
            config: Box::new(current_npmrc()),
            package_json: Box::new(PackageJson::create_if_needed(package_json_path.into())?),
            http_client: Box::new(reqwest::Client::new()),
            package_cache: PackageCache::new(),
        })
    }
}
