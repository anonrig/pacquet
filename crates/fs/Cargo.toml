[package]
name                 = "pacquet-fs"
description          = "Filesystem utility functions used by pacquet"
version              = "0.0.1"
publish              = false
authors.workspace    = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

[dependencies]
pacquet-task-queue = { workspace = true }

derive_more = { workspace = true }
miette      = { workspace = true }

[target.'cfg(windows)'.dependencies]
junction = { workspace = true }
