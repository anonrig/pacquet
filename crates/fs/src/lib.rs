mod ensure_file;
mod symlink_dir;

pub use ensure_file::*;
pub use symlink_dir::*;

pub mod file_mode;
