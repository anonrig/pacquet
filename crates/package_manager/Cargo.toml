[package]
name                  = "pacquet_package_manager"
version               = "0.0.1"
publish               = false
authors.workspace     = true
description.workspace = true
edition.workspace     = true
homepage.workspace    = true
keywords.workspace    = true
license.workspace     = true
repository.workspace  = true

[dependencies]
pacquet_fs          = { workspace = true }
pacquet_lockfile    = { workspace = true }
pacquet_npmrc       = { workspace = true }
pacquet_registry    = { workspace = true }
pacquet_tarball     = { workspace = true }

derive_more   = { workspace = true }
futures-util  = { workspace = true }
pipe-trait    = { workspace = true }
rayon         = { workspace = true }
reflink-copy  = { workspace = true }
reqwest       = { workspace = true }
tracing       = { workspace = true }
miette        = { workspace = true }

[dev-dependencies]
node-semver       = { workspace = true }
pretty_assertions = { workspace = true }
tempfile          = { workspace = true }
tokio             = { workspace = true }
