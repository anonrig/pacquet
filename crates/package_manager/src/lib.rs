mod create_cas_files;
mod create_symlink_layout;
mod create_virtual_dir_by_snapshot;
mod create_virtual_store;
mod install_package_by_snapshot;
mod install_package_from_registry;
mod link_file;
mod symlink_package;

pub use create_cas_files::{create_cas_files, CreateCasFilesError};
pub use create_symlink_layout::create_symlink_layout;
pub use create_virtual_dir_by_snapshot::{CreateVirtualDirBySnapshot, CreateVirtualDirError};
pub use create_virtual_store::CreateVirtualStore;
pub use install_package_by_snapshot::{InstallPackageBySnapshot, InstallPackageBySnapshotError};
pub use install_package_from_registry::{
    InstallPackageFromRegistry, InstallPackageFromRegistryError,
};
pub use link_file::{link_file, LinkFileError};
pub use symlink_package::{symlink_package, SymlinkPackageError};
