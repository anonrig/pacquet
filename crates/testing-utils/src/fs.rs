use std::{io, path::Path};

pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

pub fn get_all_folders(root: &std::path::Path) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_str().expect("invalid UTF-8"))
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path);
            }
        }
    }
    files.sort();
    files
}

// Helper function to check if a path is a symlink or junction
pub fn is_symlink_or_junction(path: &Path) -> io::Result<bool> {
    #[cfg(windows)]
    return junction::exists(&path);

    #[cfg(not(windows))]
    return Ok(path.is_symlink());
}
