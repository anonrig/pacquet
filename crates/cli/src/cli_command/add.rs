use crate::package_manager::{PackageManager, PackageManagerError};
use clap::Parser;
use pacquet_package_json::DependencyGroup;
use pacquet_package_manager::Install;
use pacquet_registry::{PackageTag, PackageVersion};

#[derive(Debug, Parser)]
pub struct AddDependencyOptions {
    /// Install the specified packages as regular dependencies.
    #[clap(short = 'P', long = "save-prod", group = "dependency_group")]
    save_prod: bool,
    /// Install the specified packages as devDependencies.
    #[clap(short = 'D', long = "save-dev", group = "dependency_group")]
    save_dev: bool,
    /// Install the specified packages as optionalDependencies.
    #[clap(short = 'O', long = "save-optional", group = "dependency_group")]
    save_optional: bool,
    /// Using --save-peer will add one or more packages to peerDependencies and install them as dev dependencies
    #[clap(long = "save-peer", group = "dependency_group")]
    save_peer: bool,
}

impl AddDependencyOptions {
    pub fn dependency_groups(&self) -> impl Iterator<Item = DependencyGroup> {
        let &AddDependencyOptions { save_prod, save_dev, save_optional, save_peer } = self;
        let has_prod = save_prod || (!save_dev && !save_optional && !save_peer); // no --save-* flags implies --save-prod
        let has_dev = save_dev || (!save_prod && !save_optional && save_peer); // --save-peer with nothing else implies --save-dev
        let has_optional = save_optional;
        let has_peer = save_peer;
        std::iter::empty()
            .chain(has_prod.then_some(DependencyGroup::Default))
            .chain(has_dev.then_some(DependencyGroup::Dev))
            .chain(has_optional.then_some(DependencyGroup::Optional))
            .chain(has_peer.then_some(DependencyGroup::Peer))
    }
}

#[derive(Debug, Parser)]
pub struct AddArgs {
    /// Name of the package
    pub package: String,
    /// --save-prod, --save-dev, --save-optional, --save-peer
    #[clap(flatten)]
    pub dependency_options: AddDependencyOptions,
    /// Saved dependencies will be configured with an exact version rather than using
    /// pacquet's default semver range operator.
    #[clap(short = 'E', long = "save-exact")]
    pub save_exact: bool,
    /// The directory with links to the store (default is node_modules/.pacquet).
    /// All direct and indirect dependencies of the project are linked into this directory
    #[clap(long = "virtual-store-dir", default_value = "node_modules/.pacquet")]
    pub virtual_store_dir: String,
}

impl PackageManager {
    pub async fn add(&mut self, args: &AddArgs) -> Result<(), PackageManagerError> {
        let PackageManager { config, package_json, lockfile, http_client, tarball_cache } = self;

        let latest_version = PackageVersion::fetch_from_registry(
            &args.package,
            PackageTag::Latest, // TODO: add support for specifying tags
            http_client,
            &config.registry,
        )
        .await
        .expect("resolve latest tag"); // TODO: properly propagate this error

        let version_range = latest_version.serialize(args.save_exact);
        for dependency_group in args.dependency_options.dependency_groups() {
            package_json
                .add_dependency(&args.package, &version_range, dependency_group)
                .map_err(PackageManagerError::PackageJson)?;
        }

        Install {
            tarball_cache,
            http_client,
            config,
            package_json,
            lockfile: lockfile.as_ref(),
            dependency_groups: args.dependency_options.dependency_groups(),
            frozen_lockfile: false,
        }
        .run()
        .await;

        package_json.save().map_err(PackageManagerError::PackageJson)?;

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use pacquet_npmrc::Npmrc;
    use pacquet_package_json::{DependencyGroup, PackageJson};
    use pacquet_testing_utils::fs::{get_all_folders, get_filenames_in_folder};
    use pretty_assertions::assert_eq;
    use std::{env, fs};
    use tempfile::tempdir;

    #[tokio::test]
    pub async fn should_install_all_dependencies() {
        let dir = tempdir().unwrap();
        let virtual_store_dir = dir.path().join("node_modules/.pacquet");
        let current_directory = env::current_dir().unwrap();
        env::set_current_dir(&dir).unwrap();
        let package_json = dir.path().join("package.json");
        let mut manager = PackageManager::new(&package_json, Npmrc::current().leak()).unwrap();

        // It should create a package_json if not exist
        assert!(package_json.exists());

        let args = AddArgs {
            package: "is-even".to_string(),
            dependency_options: AddDependencyOptions {
                save_prod: false,
                save_dev: false,
                save_peer: false,
                save_optional: false,
            },
            save_exact: false,
            virtual_store_dir: virtual_store_dir.to_string_lossy().to_string(),
        };
        manager.add(&args).await.unwrap();

        insta::assert_debug_snapshot!(get_all_folders(dir.path()));

        // Ensure that is-buffer does not have any dependencies
        let is_buffer_path = virtual_store_dir.join("is-buffer@1.1.6/node_modules");
        assert_eq!(get_filenames_in_folder(&is_buffer_path), vec!["is-buffer"]);

        // Ensure that is-even have correct dependencies
        let is_even_path = virtual_store_dir.join("is-even@1.0.0/node_modules");
        assert_eq!(get_filenames_in_folder(&is_even_path), vec!["is-even", "is-odd"]);

        // Ensure that is-number does not have any dependencies
        let is_number_path = virtual_store_dir.join("is-number@3.0.0/node_modules");
        assert_eq!(get_filenames_in_folder(&is_number_path), vec!["is-number", "kind-of"]);

        env::set_current_dir(&current_directory).unwrap();
    }

    #[tokio::test]
    #[cfg(not(target_os = "windows"))]
    pub async fn should_symlink_correctly() {
        let dir = tempdir().unwrap();
        let virtual_store_dir = dir.path().join("node_modules/.pacquet");
        let current_directory = env::current_dir().unwrap();
        env::set_current_dir(&dir).unwrap();
        let package_json = dir.path().join("package.json");
        let mut manager = PackageManager::new(&package_json, Npmrc::current().leak()).unwrap();

        let args = AddArgs {
            package: "is-odd".to_string(),
            dependency_options: AddDependencyOptions {
                save_prod: false,
                save_dev: false,
                save_peer: false,
                save_optional: false,
            },
            save_exact: false,
            virtual_store_dir: virtual_store_dir.to_string_lossy().to_string(),
        };
        manager.add(&args).await.unwrap();

        insta::assert_debug_snapshot!(get_all_folders(dir.path()));

        // Make sure the symlinks are correct
        assert_eq!(
            fs::read_link(virtual_store_dir.join("is-odd@3.0.1/node_modules/is-number")).unwrap(),
            fs::canonicalize(virtual_store_dir.join("is-number@6.0.0/node_modules/is-number"))
                .unwrap(),
        );
        env::set_current_dir(&current_directory).unwrap();
    }

    #[tokio::test]
    pub async fn should_add_to_package_json() {
        let dir = tempdir().unwrap();
        let virtual_store_dir = dir.path().join("node_modules/.pacquet");
        let current_directory = env::current_dir().unwrap();
        env::set_current_dir(&dir).unwrap();
        let package_json = dir.path().join("package.json");
        let mut manager = PackageManager::new(&package_json, Npmrc::current().leak()).unwrap();

        let args = AddArgs {
            package: "is-odd".to_string(),
            dependency_options: AddDependencyOptions {
                save_prod: false,
                save_dev: false,
                save_peer: false,
                save_optional: false,
            },
            save_exact: false,
            virtual_store_dir: virtual_store_dir.to_string_lossy().to_string(),
        };
        manager.add(&args).await.unwrap();
        let file = PackageJson::from_path(dir.path().join("package.json")).unwrap();
        assert!(file.dependencies([DependencyGroup::Default]).any(|(k, _)| k == "is-odd"));
        env::set_current_dir(&current_directory).unwrap();
    }

    #[tokio::test]
    pub async fn should_add_dev_dependency() {
        let dir = tempdir().unwrap();
        let virtual_store_dir = dir.path().join("node_modules/.pacquet");
        let current_directory = env::current_dir().unwrap();
        env::set_current_dir(&dir).unwrap();
        let package_json = dir.path().join("package.json");
        let mut manager = PackageManager::new(&package_json, Npmrc::current().leak()).unwrap();

        let args = AddArgs {
            package: "is-odd".to_string(),
            dependency_options: AddDependencyOptions {
                save_prod: false,
                save_dev: true,
                save_peer: false,
                save_optional: false,
            },
            save_exact: false,
            virtual_store_dir: virtual_store_dir.to_string_lossy().to_string(),
        };
        manager.add(&args).await.unwrap();
        let file = PackageJson::from_path(dir.path().join("package.json")).unwrap();
        assert!(file.dependencies([DependencyGroup::Dev]).any(|(k, _)| k == "is-odd"));
        env::set_current_dir(&current_directory).unwrap();
    }

    #[tokio::test]
    pub async fn should_add_peer_dependency() {
        let dir = tempdir().unwrap();
        let virtual_store_dir = dir.path().join("node_modules/.pacquet");
        let current_directory = env::current_dir().unwrap();
        env::set_current_dir(&dir).unwrap();
        let package_json = dir.path().join("package.json");
        let mut manager = PackageManager::new(&package_json, Npmrc::current().leak()).unwrap();

        let args = AddArgs {
            package: "is-odd".to_string(),
            dependency_options: AddDependencyOptions {
                save_prod: false,
                save_dev: false,
                save_peer: true,
                save_optional: false,
            },
            save_exact: false,
            virtual_store_dir: virtual_store_dir.to_string_lossy().to_string(),
        };
        manager.add(&args).await.unwrap();
        let file = PackageJson::from_path(dir.path().join("package.json")).unwrap();
        assert!(file.dependencies([DependencyGroup::Dev]).any(|(k, _)| k == "is-odd"));
        assert!(file.dependencies([DependencyGroup::Peer]).any(|(k, _)| k == "is-odd"));
        env::set_current_dir(&current_directory).unwrap();
    }
}
