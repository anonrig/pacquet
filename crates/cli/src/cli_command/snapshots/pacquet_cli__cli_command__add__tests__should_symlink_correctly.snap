---
source: crates/cli/src/cli_command/add.rs
assertion_line: 164
expression: get_all_folders(dir.path())
---
[
    "node_modules",
    "node_modules/.pacquet",
    "node_modules/.pacquet/is-number@6.0.0",
    "node_modules/.pacquet/is-number@6.0.0/node_modules",
    "node_modules/.pacquet/is-number@6.0.0/node_modules/is-number",
    "node_modules/.pacquet/is-odd@3.0.1",
    "node_modules/.pacquet/is-odd@3.0.1/node_modules",
    "node_modules/.pacquet/is-odd@3.0.1/node_modules/is-number",
    "node_modules/.pacquet/is-odd@3.0.1/node_modules/is-odd",
    "node_modules/is-odd",
]
